### PR TITLE
Fix #14 added latest numpy netcdf4 types

### LINKED ISSUE #14: not all types are written correctly or even written
Test/test_numpy_nc4.py

types failing for sure:

numpy.int, numpy.uint8, numpy.ubyte, etc...

### PR DESCRIPTION
**Pull request by dnadeau4**

### CODE REVIEW

**Comment by durack1** [comment]
@dnadeau3 @doutriaux1, great! I'm looking forward to testing this out!!

File: Lib/MV2.py [hunk 1]
```diff
@@ -3,8 +3,7 @@
 
 "CDMS Variable objects, MaskedArray interface"
 import numpy
-import typeconv
-from numpy import character, float, float32, float64, int, int8, int16, int32
+from numpy import character, float, float32, float64, int, int8, int16, int32, int64, byte, ubyte, uint8, uint16, uint32, uint64, long
```
Comment (doutriaux1) [inline Lib/MV2.py]: should we break this long line?

File: Lib/dataset.py [hunk 2]
```diff
@@ -1311,7 +1310,7 @@ def createAxis(self, name, ar, unlimited=0):
             typecode = ar.dtype.char
 
         # Compatibility: revert to old typecode for cdunif
-        typecode = typeconv.oldtypecodes[typecode]
+#        typecode = typeconv.oldtypecodes[typecode]
```
Comment (doutriaux1) [inline Lib/dataset.py:1313]: so we do not need convertion anymore? That's great!

File: Lib/dataset.py [hunk 3]
```diff
@@ -1789,8 +1792,8 @@ def createVariableCopy(self, var, id=None, attributes=None, axes=None, extbounds
                         attributes['_FillValue'] = numpy.array(var.missing_value).astype(var.dtype)
                         attributes['missing_value'] = numpy.array(var.missing_value).astype(var.dtype)
                 else:
-                    attributes['_FillValue'] = fill_value
-                    attributes['missing_value'] = fill_value
+                    attributes['_FillValue'] = numpy.array(fill_value).astype(var.dtype)
```
Comment (doutriaux1) [inline Lib/dataset.py:1795]: hum... did you test that with 32bit? I remember numpy complaining a lot about float32 casting

File: Lib/dataset.py [hunk 4]
```diff
@@ -1889,10 +1892,15 @@ def write(self, var, attributes=None, axes=None, extbounds=None, id=None,
...
         # Make var an AbstractVariable
         if dtype is None and typecode is not None:
-            dtype = typeconv.convtypecode2(typecode)
+#            dtype = typeconv.convtypecode2(typecode)
+            dtype = typecode
         typecode = dtype
         if typecode is not None and var.dtype.char != typecode:
             var = var.astype(typecode)
+        if var.dtype.char == 'l' and Cdunif.CdunifGetNCFLAGS("classic"):
+            var = var.astype(numpy.int32)
```
Comment (doutriaux1) [inline Lib/dataset.py:1901]: test for this?

File: Lib/tvariable.py [hunk 5]
```diff
@@ -164,10 +163,12 @@ def __init__(self, data, typecode=None, copy=1, savespace=0,
 
         # tile index, None means no mosaic
         self.tileIndex = None
-
+        import pdb
```
Comment (doutriaux1) [inline Lib/tvariable.py]: oooopssy....

File: tests/test_allMIPs.py [hunk 6]
```diff
@@ -16,6 +16,8 @@ def tearDown(self):
         os.remove(self.filename)
 
     def testinput4MIPs(self):
+        import pdb
```
Comment (doutriaux1) [inline tests/test_allMIPs.py]: ooopppsy again

File: tests/test_allMIPs.py [hunk 7]
```diff
@@ -16,6 +16,8 @@ def tearDown(self):
...
+        pdb.set_trace()
```
Comment (doutriaux1) [inline tests/test_allMIPs.py]: oohhh.....

**Review by doutriaux1 — COMMENTED** [review]
@dnadeau4 do we have a test that writes data in every type?
Would be nice to see if can still down/up cast at write time as well i.e write a double as float32 or int as int32

**Comment by doutriaux1** [comment]
why close?

**Comment by dnadeau4** [comment]
To early to revise, I will reopen when ready to merge.

File: Lib/MV2.py (hunk 1 above)
Comment (dnadeau4) [reply Lib/MV2.py]: I will pep8 when ready.

File: Lib/dataset.py (hunk 2 above)
Comment (dnadeau4) [reply Lib/dataset.py:1313]: :+1:

File: Lib/dataset.py (hunk 3 above)
Comment (dnadeau4) [reply Lib/dataset.py:1795]: I tested everything.  :smirk:

File: tests/test_allMIPs.py (hunk 7 above)
Comment (dnadeau4) [reply tests/test_allMIPs.py]: That's why I closed it.

File: Lib/dataset.py (hunk 4 above)
Comment (dnadeau4) [reply Lib/dataset.py:1901]: This is important for the attributes.  I cannot write int64 attribute in a `netCDF classic format`.   Data and _FillValue needs to be int64 but not `integer` attributes in this case.
Yes test fails if this is not there!

Comment (dnadeau4) [reply Lib/dataset.py:1901]: Actually `l` stand for long. Numpy `LONG` is `int64` and `NC_LONG` is `int32`.

File: Lib/dataset.py (hunk 3 above)
Comment (dnadeau4) [reply Lib/dataset.py:1795]: https://github.com/UV-CDAT/cdms/blob/issue%2314_types/tests/test_numpy_nc4.py

**Pull request merged**